### PR TITLE
Generalised description edit flow

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
- Tap on any description to change it based on the permission
- Three dot menu option will also have Edit Description menu based on the permission

Side-note:
- On Appflowy HTML editor, default auto focus is set but it is always setting cursor at the first position. Don't know where this is issue from Appflowy or we need to manage something from our side. 

fixes, #1842 

1. Space About:

https://github.com/acterglobal/a3/assets/51526480/2cb642d0-ff3f-4cd0-aa90-da3e88949ef5

2. Chat Topic:

https://github.com/acterglobal/a3/assets/51526480/04511be1-2baa-44dc-805a-d4616b52e93e

3. Event Description:

https://github.com/acterglobal/a3/assets/51526480/584695c4-2884-4560-a46d-4550ae8fbe21

4. Pin Description:

https://github.com/acterglobal/a3/assets/51526480/17662399-f5c5-4344-8cfb-ff92bdcacaa7

5. Task Item Description:

https://github.com/acterglobal/a3/assets/51526480/c9c13be9-4822-4208-b724-a6c5b3b2ba8c

6. Task List Description:

https://github.com/acterglobal/a3/assets/51526480/1350cd1f-3560-4b2f-9428-580ac09f31ef

### CODE REVIEW

File: app/lib/features/tasks/pages/task_item_detail_page.dart [hunk 1]
```diff
@@ -152,6 +166,54 @@ class TaskItemDetailPage extends ConsumerWidget {
...
+      if (context.mounted) context.pop();
+    } catch (e, st) {
+      _log.severe('Failed to update event description', e, st);
+      EasyLoading.dismiss();
+      if (!context.mounted) return;
+      EasyLoading.showError(
+        L10n.of(context).errorUpdatingDescription(e),
+        duration: const Duration(seconds: 3),
+      );
+    }
```
Comment (gnunicorn) [inline app/lib/features/tasks/pages/task_item_detail_page.dart:214]: doing this pattern so often, it makes you wonder if we want to generalize this ... (but maybe we want to do that later)

File: app/lib/features/tasks/pages/task_list_details_page.dart [hunk 2]
```diff
@@ -88,9 +113,19 @@ class _TaskListPageState extends ConsumerState<TaskListDetailPage> {
...
-        Text(
-          taskListData.description()!.body(),
-          style: Theme.of(context).textTheme.labelMedium,
+        GestureDetector(
+          onTap: () {
+            showEditDescriptionSheet(taskListData);
+          },
+          child: taskListData.description()!.formattedBody() != null
+              ? RenderHtml(
+                  text: taskListData.description()!.formattedBody()!,
```
Comment (gnunicorn) [inline app/lib/features/tasks/pages/task_list_details_page.dart:122]: see above. also the idea to just make that a Widget itself given a `MsgContent` as input...?

**Review by gnunicorn — APPROVED** [review]
Looking good. Some remarks but none of them blocking a review.

File: app/lib/common/widgets/edit_html_description_sheet.dart [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            ),
+            child: HtmlEditor(
+              editorState: textEditorState,
+              editable: true,
+              autoFocus: true,
+              onChanged: (body, html) {
+                final document = html != null
+                    ? ActerDocumentHelpers.fromHtml(html)
+                    : ActerDocumentHelpers.fromMarkdown(body);
+                textEditorState = EditorState(document: document);
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/common/widgets/edit_html_description_sheet.dart]: I just copied it from previous implementation but you right, it doesn't required at all.

File: app/lib/common/widgets/edit_html_description_sheet.dart [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                  final htmlBodyDescription = textEditorState.intoHtml();
+                  final plainDescription = textEditorState.intoMarkdown();
+                  if (htmlBodyDescription == widget.descriptionHtmlValue ||
+                      plainDescription == widget.descriptionMarkdownValue) {
+                    context.pop();
+                    return;
+                  }
+
+                  // Need to update change of tile
+                  widget.onSave(textEditorState);
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/common/widgets/edit_html_description_sheet.dart]: Nice suggestion

File: app/lib/features/pins/widgets/pin_item.dart [hunk 5]
```diff
@@ -150,66 +155,109 @@ class _PinDescriptionWidgetConsumerState
...
+                    final document = html != null
+                        ? ActerDocumentHelpers.fromHtml(html)
+                        : ActerDocumentHelpers.fromMarkdown(body);
+                    textEditorState = EditorState(document: document);
+                  } else {
+                    textEditorState = EditorState(
+                      document: ActerDocumentHelpers.fromMarkdown(body),
+                    );
+                  }
+                },
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/pins/widgets/pin_item.dart:189]: For pins old work, I would like to do it on separate PR. There are many of the code can be refactor and removed after new inline approach.

File: app/lib/features/pins/widgets/pin_item.dart [hunk 6]
```diff
@@ -150,66 +155,109 @@ class _PinDescriptionWidgetConsumerState
...
+            onCancel: () {
+              final content = widget.pin.content();
+              if (content != null) {
+                textEditorState = EditorState(
+                  document: ActerDocumentHelpers.fromMsgContent(content),
+                );
+              } else {
+                textEditorState = EditorState.blank();
+              }
+              pinEditNotifier.setEditMode(false);
```
[the remark this answers is not in the document]
Comment (kumarpalsinh25) [reply app/lib/features/pins/widgets/pin_item.dart:204]: As said on above comment, I would prefer to do pins refactor on separate PR. Can't remove entire edit mode as of now as there is edit link field still stick with this page. We will get rid of all this during pins refactoring. :)

File: app/lib/features/tasks/pages/task_item_detail_page.dart (hunk 1 above)
Comment (kumarpalsinh25) [reply app/lib/features/tasks/pages/task_item_detail_page.dart:213]: Yes true, I was also thinking of the same during this code writing. 
Planning to first later of generalisation and then do this type of second layer of generalisation. :)

**Pull request merged**